File: releasing/plugin-commands-publishing/src/publish.ts
import path from 'path'
import { docsUrl, readProjectManifest } from '@pnpm/cli-utils'
import { FILTERING } from '@pnpm/common-cli-options-help'
import { type Config, types as allTypes } from '@pnpm/config'
import { PnpmError } from '@pnpm/error'
import { runLifecycleHook, type RunLifecycleHookOptions } from '@pnpm/lifecycle'
import { type ProjectManifest } from '@pnpm/types'
import { getCurrentBranch, isGitRepo, isRemoteHistoryClean, isWorkingTreeClean } from '@pnpm/git-utils'
import enquirer from 'enquirer'
import rimraf from '@zkochan/rimraf'
import { pick } from 'ramda'
import realpathMissing from 'realpath-missing'
import renderHelp from 'render-help'
import { temporaryDirectory } from 'tempy'
import { extractManifestFromPacked, isTarballPath } from './extractManifestFromPacked.js'
import { optionsWithOtpEnv } from './otpEnv.js'
import * as pack from './pack.js'
import { publishPackedPkg } from './publishPackedPkg.js'
import { recursivePublish, type PublishRecursiveOpts } from './recursivePublish.js'

export function rcOptionsTypes (): Record<string, unknown> {
  return pick([
    'access',
    'git-checks',
    'ignore-scripts',
    'provenance',
    'npm-path',
    'otp',
    'publish-branch',
    'registry',
    'tag',
    'unsafe-perm',
    'embed-readme',
  ], allTypes)
}

export function cliOptionsTypes (): Record<string, unknown> {
  return {
    ...rcOptionsTypes(),
    'dry-run': Boolean,
    force: Boolean,
    json: Boolean,
    otp: String,
    recursive: Boolean,
    'report-summary': Boolean,
  }
}

export const commandNames = ['publish']

export function help (): string {
  return renderHelp({
    description: 'Publishes a package to the npm registry.',
    descriptionLists: [
      {
        title: 'Options',

        list: [
          {
            description: "Don't check if current branch is your publish branch, clean, and up to date",
            name: '--no-git-checks',
          },
          {
            description: 'Sets branch name to publish. Default is master',
            name: '--publish-branch',
          },
          {
            description: 'Does everything a publish would do except actually publishing to the registry',
            name: '--dry-run',
          },
          {
            description: 'Show information in JSON format',
            name: '--json',
          },
          {
            description: 'Registers the published package with the given tag. By default, the "latest" tag is used.',
            name: '--tag <tag>',
          },
          {
            description: 'Tells the registry whether this package should be published as public or restricted',
            name: '--access <public|restricted>',
          },
          {
            description: 'Ignores any publish related lifecycle scripts (prepublishOnly, postpublish, and the like)',
            name: '--ignore-scripts',
          },
          {
            description: 'Packages are proceeded to be published even if their current version is already in the registry. This is useful when a "prepublishOnly" script bumps the version of the package before it is published',
            name: '--force',
          },
          {
            description: 'Save the list of the newly published packages to "pnpm-publish-summary.json". Useful when some other tooling is used to report the list of published packages.',
            name: '--report-summary',
          },
          {
            description: 'When publishing packages that require two-factor authentication, this option can specify a one-time password',
            name: '--otp',
          },
          {
            description: 'Publish all packages from the workspace',
            name: '--recursive',
            shortAlias: '-r',
          },
        ],
      },
      FILTERING,
    ],
    url: docsUrl('publish'),
    usages: ['pnpm publish [<tarball>|<dir>] [--tag <tag>] [--access <public|restricted>] [options]'],
  })
}

const GIT_CHECKS_HINT = 'If you want to disable Git checks on publish, set the "git-checks" setting to "false", or run again with "--no-git-checks".'

export async function handler (
  opts: Omit<PublishRecursiveOpts, 'workspaceDir'> & {
    argv: {
      original: string[]
    }
    engineStrict?: boolean
    recursive?: boolean
    workspaceDir?: string
  } & Pick<Config, 'allProjects' | 'bin' | 'gitChecks' | 'ignoreScripts' | 'pnpmHomeDir' | 'publishBranch' | 'embedReadme'>,
  params: string[]
): Promise<{ exitCode?: number } | undefined> {
  const result = await publish(opts, params)
  if (result?.manifest) return
  return result
}

export interface PublishResult {
  exitCode?: number
  manifest?: ProjectManifest
}

export async function publish (
  opts: Omit<PublishRecursiveOpts, 'workspaceDir'> & {
    argv: {
      original: string[]
    }
    engineStrict?: boolean
    recursive?: boolean
    workspaceDir?: string
  } & Pick<Config, 'allProjects' | 'bin' | 'gitChecks' | 'ignoreScripts' | 'pnpmHomeDir' | 'publishBranch' | 'embedReadme' | 'packGzipLevel'>,
  params: string[]
): Promise<PublishResult> {
  if (opts.gitChecks !== false && await isGitRepo()) {
    if (!(await isWorkingTreeClean())) {
      throw new PnpmError('GIT_UNCLEAN', 'Unclean working tree. Commit or stash changes first.', {
        hint: GIT_CHECKS_HINT,
      })
    }
    const branches = opts.publishBranch ? [opts.publishBranch] : ['master', 'main']
    const currentBranch = await getCurrentBranch()
    if (currentBranch === null) {
      throw new PnpmError(
        'GIT_UNKNOWN_BRANCH',
        `The Git HEAD may not attached to any branch, but your "publish-branch" is set to "${branches.join('|')}".`,
        {
          hint: GIT_CHECKS_HINT,
        }
      )
    }
    if (!branches.includes(currentBranch)) {
      const { confirm } = await enquirer.prompt({
        message: `You're on branch "${currentBranch}" but your "publish-branch" is set to "${branches.join('|')}". \
Do you want to continue?`,
        name: 'confirm',
        type: 'confirm',
      } as any) as any // eslint-disable-line @typescript-eslint/no-explicit-any

      if (!confirm) {
        throw new PnpmError('GIT_NOT_CORRECT_BRANCH', `Branch is not on '${branches.join('|')}'.`, {
          hint: GIT_CHECKS_HINT,
        })
      }
    }
    if (!(await isRemoteHistoryClean())) {
      throw new PnpmError('GIT_NOT_LATEST', 'Remote history differs. Please pull changes.', {
        hint: GIT_CHECKS_HINT,
      })
    }
  }
  if (opts.recursive && (opts.selectedProjectsGraph != null)) {
    const { exitCode } = await recursivePublish({
      ...opts,
      selectedProjectsGraph: opts.selectedProjectsGraph,
      workspaceDir: opts.workspaceDir ?? process.cwd(),
    })
    return { exitCode }
  }

  opts = optionsWithOtpEnv(opts, process.env)

  const dirInParams = (params.length > 0) ? params[0] : undefined

  if (dirInParams != null && isTarballPath(dirInParams)) {
    const tarballPath = dirInParams
    const publishedManifest = await extractManifestFromPacked(tarballPath)
    await publishPackedPkg({
      tarballPath,
      publishedManifest,
    }, opts)
    return { exitCode: 0 }
  }

  const dir = dirInParams ?? opts.dir ?? process.cwd()

  const _runScriptsIfPresent = runScriptsIfPresent.bind(null, {
    depPath: dir,
    extraBinPaths: opts.extraBinPaths,
    extraEnv: opts.extraEnv,
    pkgRoot: dir,
    rawConfig: opts.rawConfig,
    rootModulesDir: await realpathMissing(path.join(dir, 'node_modules')),
    stdio: 'inherit',
    unsafePerm: true, // when running scripts explicitly, assume that they're trusted.
  })
  const { manifest } = await readProjectManifest(dir, opts)
  // Unfortunately, we cannot support postpack at the moment
  if (!opts.ignoreScripts) {
    await _runScriptsIfPresent([
      'prepublishOnly',
      'prepublish',
    ], manifest)
  }

  // We have to publish the tarball from another location.
  // Otherwise, npm would publish the package with the package.json file
  // from the current working directory, ignoring the package.json file
  // that was generated and packed to the tarball.
  const packDestination = temporaryDirectory()
  try {
    const packResult = await pack.api({
      ...opts,
      dir,
      packDestination,
      dryRun: false,
    })
    await publishPackedPkg(packResult, opts)
  } finally {
    await rimraf(packDestination)
  }

  if (!opts.ignoreScripts) {
    await _runScriptsIfPresent([
      'publish',
      'postpublish',
    ], manifest)
  }
  return { manifest }
}

export async function runScriptsIfPresent (
  opts: RunLifecycleHookOptions,
  scriptNames: string[],
  manifest: ProjectManifest
): Promise<void> {
  for (const scriptName of scriptNames) {
    if (!manifest.scripts?.[scriptName]) continue
    await runLifecycleHook(scriptName, manifest, opts) // eslint-disable-line no-await-in-loop
  }
}

File: pacquet/crates/cli/src/cli_args/publish.rs
//! `pacquet publish` — publish a package to an npm registry.
//!
//! Ports pnpm's
//! [`publish` command](https://github.com/pnpm/pnpm/blob/54c5c0e028/pnpm11/releasing/commands/src/publish/publish.ts).
//! The registry-facing work (OIDC, OTP, the publish document and PUT) lives in
//! [`pacquet_publish`]; this module maps the resolved [`Config`] and CLI flags
//! onto its options, runs the git checks and publish-lifecycle scripts, and
//! packs the project before handing the tarball off.
//!
//! Scope versus upstream: `--recursive` / `--batch` (workspace and batched
//! publishing) are accepted for surface parity but not yet ported — they error
//! rather than silently doing nothing.

use std::{collections::HashMap, path::Path};

use clap::Args;
use miette::{Context, IntoDiagnostic};
use pacquet_config::Config;
use pacquet_executor::{RunPostinstallHooks, ScriptsPrependNodePath, run_lifecycle_hook};
use pacquet_pack::{Host as PackHost, PackOptions, PackResult, api as pack_api};
use pacquet_publish::{
    Access, Host, OidcHttpOptions, PackedPkg, PublishNetwork, PublishPackedPkgOptions,
    extract_manifest_from_packed, is_tarball_path, publish_packed_pkg, resolve_otp_from_env,
    run_git_checks,
};
use pacquet_reporter::Reporter;
use serde_json::Value;

use crate::cli_args::registry_client::build_registry_client;

/// `pacquet publish` arguments. The `-r` / `--recursive` selector is a global
/// flag on [`crate::CliArgs`] and threaded into [`Self::run`].
#[derive(Debug, Args)]
pub struct PublishArgs {
    /// Tarball or directory to publish. Defaults to the current directory.
    pub package: Option<String>,

    /// Do everything `publish` would do except uploading to the registry.
    #[clap(long)]
    pub dry_run: bool,

    /// Print the per-package publish summary in JSON.
    #[clap(long)]
    pub json: bool,

    /// Register the published package under this tag instead of `latest`.
    #[clap(long)]
    pub tag: Option<String>,

    /// Publish the package as `public` or `restricted`.
    #[clap(long, value_parser = ["public", "restricted"])]
    pub access: Option<String>,

    /// Generate a provenance attestation for the published package.
    #[clap(long)]
    pub provenance: bool,

    /// Don't run publish-related lifecycle scripts.
    #[clap(long = "ignore-scripts")]
    pub ignore_scripts: bool,

    /// Keep the original `packageManager` field and publish-lifecycle scripts
    /// in the published manifest instead of stripping them.
    #[clap(long = "skip-manifest-obfuscation")]
    pub skip_manifest_obfuscation: bool,

    /// One-time password for two-factor-authenticated registries.
    #[clap(long)]
    pub otp: Option<String>,

    /// The branch publishing is allowed from. Defaults to `master` / `main`.
    #[clap(long = "publish-branch")]
    pub publish_branch: Option<String>,

    /// Skip the git working-tree / branch / remote checks.
    #[clap(long = "no-git-checks")]
    pub no_git_checks: bool,

    /// Publish even if the version is already in the registry.
    #[clap(long)]
    pub force: bool,

    /// Publish all packages from the workspace.
    #[clap(long)]
    pub recursive: bool,

    /// Send all workspace packages in a single request (requires `--recursive`).
    #[clap(long)]
    pub batch: bool,
}

impl PublishArgs {
    /// Publish the package at `dir` (or the given tarball/directory),
    /// returning nothing — output is printed here. Ports pnpm's `publish`
    /// handler for the single-package and tarball paths.
    pub async fn run<Reporter: self::Reporter>(
        self,
        dir: &Path,
        config: &Config,
        recursive: bool,
    ) -> miette::Result<()> {
        if self.batch && !(recursive || self.recursive) {
            return Err(miette::miette!(
                code = "ERR_PNPM_BATCH_PUBLISH_REQUIRES_RECURSIVE",
                "--batch can only be used together with --recursive",
            ));
        }

        let publish_branch = self.publish_branch.as_deref();
        run_git_checks::<Host>(dir, !self.no_git_checks, publish_branch)?;

        if recursive || self.recursive {
            return Err(miette::miette!(
                code = "ERR_PNPM_RECURSIVE_PUBLISH_UNSUPPORTED",
                help =
                    "Run `pnpm publish` in each package directory, or use the TypeScript pnpm CLI.",
                "Recursive publishing is not yet supported by pacquet",
            ));
        }

        let otp = resolve_otp_from_env::<Host>(self.otp.clone());
        let opts = self.publish_options(config, otp);
        let http_client = build_registry_client(config)?;
        let network = PublishNetwork { client: &http_client, auth_headers: &config.auth_headers };

        let summary_json =
            if let Some(package) = self.package.as_deref().filter(|p| is_tarball_path(p)) {
                self.publish_tarball::<Reporter>(package, &opts, &network).await?
            } else {
                self.publish_directory::<Reporter>(
                    self.package.as_deref().unwrap_or_else(|| dir_str(dir)),
                    config,
                    &opts,
                    &network,
                )
                .await?
            };

        if self.json {
            println!("{summary_json}");
        }
        Ok(())
    }

    /// Publish a pre-built tarball: extract its manifest and hand the bytes
    /// straight to the registry (no file listing or unpacked size).
    async fn publish_tarball<Reporter: self::Reporter>(
        &self,
        tarball_path: &str,
        opts: &PublishPackedPkgOptions,
        network: &PublishNetwork<'_>,
    ) -> miette::Result<String> {
        let manifest = extract_manifest_from_packed(tarball_path)?;
        let tarball_data = std::fs::read(tarball_path)
            .into_diagnostic()
            .wrap_err_with(|| format!("read tarball {tarball_path}"))?;
        let summary = publish_packed_pkg::<Host, Reporter>(
            &PackedPkg {
                published_manifest: &manifest,
                tarball_data: &tarball_data,
                tarball_path,
                contents: &[],
                unpacked_size: 0,
            },
            opts,
            network,
        )
        .await?;
        serde_json::to_string_pretty(&summary).into_diagnostic()
    }

    /// Publish a project directory: run `prepublishOnly` / `prepublish`, pack
    /// the project into a temporary directory, publish the tarball, then run
    /// `publish` / `postpublish`.
    async fn publish_directory<Reporter: self::Reporter>(
        &self,
        dir: &str,
        config: &Config,
        opts: &PublishPackedPkgOptions,
        network: &PublishNetwork<'_>,
    ) -> miette::Result<String> {
        let project_dir = Path::new(dir);
        let manifest = pacquet_package_manifest::safe_read_package_json_from_dir(project_dir)
            .into_diagnostic()
            .wrap_err("read package.json")?
            .ok_or_else(|| {
                miette::miette!(
                    code = "ERR_PNPM_NO_IMPORTER_MANIFEST_FOUND",
                    "No package.json found in {dir}",
                )
            })?;

        if !self.ignore_scripts {
            run_publish_scripts::<Reporter>(
                project_dir,
                config,
                &manifest,
                &["prepublishOnly", "prepublish"],
            )?;
        }

        let pack_destination = tempfile::tempdir().into_diagnostic().wrap_err("create temp dir")?;
        let pack_result =
            self.pack_for_publish::<Reporter>(project_dir, config, pack_destination.path())?;
        let tarball_data = std::fs::read(&pack_result.tarball_path)
            .into_diagnostic()
            .wrap_err("read packed tarball")?;

        let summary = publish_packed_pkg::<Host, Reporter>(
            &PackedPkg {
                published_manifest: &pack_result.published_manifest,
                tarball_data: &tarball_data,
                tarball_path: &pack_result.tarball_path,
                contents: &pack_result.contents,
                unpacked_size: pack_result.unpacked_size,
            },
            opts,
            network,
        )
        .await?;
        drop(pack_destination);

        if !self.ignore_scripts {
            run_publish_scripts::<Reporter>(
                project_dir,
                config,
                &manifest,
                &["publish", "postpublish"],
            )?;
        }
        serde_json::to_string_pretty(&summary).into_diagnostic()
    }

    /// Pack the project into `pack_destination` for publishing (never a dry
    /// run; the publish itself honors `--dry-run`).
    fn pack_for_publish<Reporter: self::Reporter>(
        &self,
        dir: &Path,
        config: &Config,
        pack_destination: &Path,
    ) -> miette::Result<PackResult> {
        let options = PackOptions {
            dir: dir.to_path_buf(),
            catalogs: config.catalogs.clone().unwrap_or_default(),
            ignore_scripts: config.ignore_scripts,
            unsafe_perm: config.unsafe_perm,
            embed_readme: false,
            pack_gzip_level: None,
            node_linker: config.node_linker,
            skip_manifest_obfuscation: self.skip_manifest_obfuscation,
            user_agent: config.user_agent.clone(),
            extra_bin_paths: config.extra_bin_paths.clone(),
            extra_env: HashMap::new(),
            workspace_dir: config.workspace_dir.clone(),
            dry_run: false,
            out: None,
            pack_destination: Some(pack_destination.to_string_lossy().into_owned()),
        };
        pack_api::<Reporter, PackHost>(&options)
            .map_err(miette::Report::new)
            .wrap_err("pack the package")
    }

    /// Map the CLI flags and resolved [`Config`] onto the publish options.
    fn publish_options(&self, config: &Config, otp: Option<String>) -> PublishPackedPkgOptions {
        PublishPackedPkgOptions {
            default_registry: config.registry.clone(),
            scoped_registries: config.registries.clone(),
            access: self.access.as_deref().and_then(Access::parse),
            tag: self.tag.clone().unwrap_or_else(|| "latest".to_owned()),
            otp,
            // An absent `--provenance` leaves the decision to the OIDC flow.
            provenance: self.provenance.then_some(true),
            dry_run: self.dry_run,
            stage: false,
            http: OidcHttpOptions {
                fetch_retries: Some(config.fetch_retries),
                fetch_retry_factor: Some(f64::from(config.fetch_retry_factor)),
                fetch_retry_maxtimeout: Some(config.fetch_retry_maxtimeout),
                fetch_retry_mintimeout: Some(config.fetch_retry_mintimeout),
                fetch_timeout: Some(config.fetch_timeout),
            },
        }
    }
}

/// Run the publish-lifecycle scripts the manifest declares, in order, with
/// `unsafe_perm` (publish scripts are run explicitly and assumed trusted).
/// Mirrors pnpm's `runScriptsIfPresent`.
fn run_publish_scripts<Reporter: self::Reporter>(
    dir: &Path,
    config: &Config,
    manifest: &Value,
    script_names: &[&str],
) -> miette::Result<()> {
    let scripts = manifest.get("scripts");
    let declares = |name: &str| {
        scripts
            .and_then(|scripts| scripts.get(name))
            .and_then(Value::as_str)
            .filter(|s| !s.is_empty())
    };
    if !script_names.iter().any(|name| declares(name).is_some()) {
        return Ok(());
    }

    let dep_path = dir.to_string_lossy().into_owned();
    let root_modules_dir = dir.join("node_modules");
    let run_opts = RunPostinstallHooks {
        dep_path: &dep_path,
        pkg_root: dir,
        root_modules_dir: &root_modules_dir,
        init_cwd: dir,
        extra_bin_paths: &config.extra_bin_paths,
        extra_env: &HashMap::new(),
        node_execpath: None,
        npm_execpath: None,
        node_gyp_path: None,
        user_agent: Some(&config.user_agent),
        unsafe_perm: true,
        node_gyp_bin: None,
        scripts_prepend_node_path: ScriptsPrependNodePath::default(),
        script_shell: None,
        optional: false,
    };
    let parent_env: HashMap<String, String> = std::env::vars().collect();

    for &name in script_names {
        let Some(script) = declares(name) else { continue };
        run_lifecycle_hook::<Reporter>(name, script, &run_opts, manifest, &parent_env)
            .map_err(miette::Report::new)
            .wrap_err_with(|| format!("run the {name} script"))?;
    }
    Ok(())
}

/// Borrow `dir` as a `&str` for the default-directory case.
fn dir_str(dir: &Path) -> &str {
    dir.to_str().unwrap_or(".")
}
